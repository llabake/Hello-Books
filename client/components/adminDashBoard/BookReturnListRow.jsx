import React, { Component } from 'react';
import { connect } from 'react-redux';
import moment from "moment";
import { acceptBookReturnRequest } from '../../actions/adminAction';


/**
 * 
 * 
 * @class BookReturnListRow
 * @extends {Component}
 */
export class BookReturnListRow extends Component {
  /**
   * Creates an instance of BookReturnListRow.
   * @param {any} props 
   * @memberof BookReturnListRow
   */
  constructor(props) {
    super(props);

    this.handleReturn = this.handleReturn.bind(this)

  }

  /**
   * @returns {Object} response object
   * 
   * @param {any} event 
   * @memberof BookReturnListRow
   */
  handleReturn(event) {
    event.preventDefault();
    this.props.acceptBookReturnRequest(this.props.borrowedBook.book.id, this.props.borrowedBook.user.id)
  }

  /**
   * 
   * 
   * @returns {array} array of pending borroed books
   * @memberof BookReturnListRow
   */
  render () {
    const { borrowedBook, index } = this.props
    return (
      <tr>
        <td>{index + 1}</td>
        <td>{borrowedBook.book.title}</td>
        <td>{borrowedBook.book.author}</td>
        <td>{moment(borrowedBook.borrowDate).format('ll')}</td>
        <td>{borrowedBook.user.username}</td>
        <td id='return-book'>
          <a onClick = {this.handleReturn}
          style={{ cursor: "pointer" }}>
          <i className="material-icons">
            assignment_returned
          </i>
          </a>
        </td>
      </tr>
    )
  }
}

export const mapStateToProps = (state) => {
  return {
    errors: state.adminReducer.errors,

  }
}

export const mapDispatchToProps = dispatch => {
  return {
    acceptBookReturnRequest: (bookId, userId) => dispatch(acceptBookReturnRequest(bookId, userId)),
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(BookReturnListRow)