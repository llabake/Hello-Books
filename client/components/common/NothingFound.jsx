import React from 'react';
import PropTypes from 'prop-types';

const NothingFound = ({ text, id }) => {
  return (
    <div
      className="card-panel row center-align"
      id={id}>
      <h4>
        {text}
      </h4>
    </div>
  )
};

NothingFound.propTypes = {
  id: PropTypes.string,
  text: PropTypes.string
};

export default NothingFound;