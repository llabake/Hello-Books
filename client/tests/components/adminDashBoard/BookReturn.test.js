import React from 'react';

import {
  BookReturn,
  mapStateToProps,
  mapDispatchToProps
} from '../../../components/adminDashBoard/BookReturn';

import BookReturnListRow from '../../../components/adminDashBoard/BookReturnListRow'
import adminData from '../../mocks/userData';
import userData from '../../mocks/userData';
import NothingFound from '../../../components/common/NothingFound';
import { Pagination } from 'react-materialize';


function setup() {
  const props = {
    dispatch: jest.fn(),
    user: adminData.admin,
    errors: {},
    handleSelectedPage: jest.fn(),
    loading: false,
    returnedBookCount: 0,
    borrowedBooks: [
      {
        "message": "Borrowed books retrieved successfully",
        "borrowedBooks": [
          {
            "id": 15,
            "borrowDate": "2018-03-25T16:41:57.255Z",
            "expectedReturnDate": "2018-04-08T16:41:57.255Z",
            "actualReturnDate": null,
            "borrowStatus": "accepted",
            "returnStatus": "pending",
            "bookId": 2,
            "userId": 2,
            "user": {
              "username": "mama",
              "id": 2
            },
            "book": {
              "id": 2,
              "title": "dont be sad ",
              "author": "Aaidh ibn Abdullah al-Qarni",
              "quantity": 1,
              "borrowCount": 6
            }
          }
        ]
      }
    ],
    pendingAcceptReturnRequest: jest.fn(),
  };

  const state = {
    maxItems: 2,
    showPagination: false,
    activePage: 1,

  };
  const wrapper = shallow(<BookReturn {...props} {...state} />);

  return {
    wrapper,
    props,
    state,
  };
}

describe('The Book borrow tab', () => {
  test('should render and not crash', () => {
    const { wrapper } = setup();
    expect(wrapper).toMatchSnapshot();
  })

  test('should call componentDidMount', () => {
    let spy;
    afterEach(() => {
      spy.mockClear();
    });
    spy = jest.spyOn(BookReturn.prototype, 'componentDidMount')
    const { wrapper, props, state } = setup();
    expect(BookReturn.prototype.componentDidMount).toHaveBeenCalledTimes(1)
  })

  test('should render the book borrowed list', () => {
    const { wrapper, props, state } = setup();
    expect(wrapper.find(BookReturnListRow).length).toBe(1);
    expect(wrapper.find(Pagination).length).toBe(0);
  })

  test('should render nothing found component', () => {
    const props = {
      dispatch: jest.fn(),
      user: adminData.admin,
      errors: {},
      handleSelectedPage: jest.fn(),
      loading: false,
      returnedBookCount: 0,
      borrowedBooks: [],
      pendingAcceptReturnRequest: jest.fn(),
    }
    const wrapper = shallow(<BookReturn {...props} />);

    expect(wrapper.find(NothingFound).length).toBe(1)

  })

  test('should call handle selected page', () => {
    const { wrapper, props, state } = setup();
    const page = 2
    wrapper.setState({
      activePage: page,
    });
    wrapper.instance().handleSelectedPage(page)
    expect(wrapper.instance().state.activePage).toBe(2);
  });

  test('should call componentWillReceiveProps', () => {
    let spy;
    afterEach(() => {
      spy.mockClear();
    });

    const props = {
      dispatch: jest.fn(),
      user: adminData.admin,
      errors: {},
      handleSelectedPage: jest.fn(),
      loading: false,
      returnedBookCount: 0,
      borrowedBooks: [],
      pendingAcceptReturnRequest: jest.fn(),
    }

    const nextProps = {
      dispatch: jest.fn(),
      user: adminData.admin,
      errors: {},
      handleSelectedPage: jest.fn(),
      loading: false,
      returnedBookCount: 1,
      borrowedBooks: [
        {
          "message": "Borrowed books retrieved successfully",
          "borrowedBooks": [
            {
              "id": 15,
              "borrowDate": "2018-03-25T16:41:57.255Z",
              "expectedReturnDate": "2018-04-08T16:41:57.255Z",
              "actualReturnDate": null,
              "borrowStatus": "accepted",
              "returnStatus": "pending",
              "bookId": 2,
              "userId": 2,
              "user": {
                "username": "mama",
                "id": 2
              },
              "book": {
                "id": 2,
                "title": "dont be sad ",
                "author": "Aaidh ibn Abdullah al-Qarni",
                "quantity": 1,
                "borrowCount": 6
              }
            }
          ]
        }
      ],
      pendingAcceptReturnRequest: jest.fn(),
    }
    spy = jest.spyOn(BookReturn.prototype, 'componentWillReceiveProps');
    const wrapper = shallow(<BookReturn {...props} />)
    wrapper.setProps(nextProps);
    expect(BookReturn.prototype.componentWillReceiveProps).toHaveBeenCalledTimes(1)
    expect(spy).toHaveBeenCalled();
  })

  test('mapDispatchToProps', () => {
    const dispatch = jest.fn();

    expect(mapDispatchToProps(dispatch)).toHaveProperty('pendingAcceptReturnRequest');

    const { pendingAcceptReturnRequest } = mapDispatchToProps(dispatch);
    pendingAcceptReturnRequest();
    expect(dispatch).toHaveBeenCalled();
  });

  test('mapStateToProps', () => {
    expect(mapStateToProps({ adminReducer: {} })).toHaveProperty('borrowedBooks');
    expect(mapStateToProps({ adminReducer: {} })).toHaveProperty('loading');
    expect(mapStateToProps({ adminReducer: {} })).toHaveProperty('returnedBookCount');
  });

})