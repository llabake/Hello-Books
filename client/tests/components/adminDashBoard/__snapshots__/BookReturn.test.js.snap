// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The Book borrow tab should render and not crash 1`] = `
<div
  id="return"
>
  <div
    className="col s12"
  >
    <div
      className="card-panel"
    >
      <table
        className="bordered centered highlight responsive-table"
      >
        <thead>
          <tr>
            <th>
              S/N
            </th>
            <th>
              Title
            </th>
            <th>
              Author 
            </th>
            <th>
              Date Borrowed
            </th>
            <th>
              User
            </th>
            <th>
              Return
            </th>
          </tr>
        </thead>
        <tbody>
          <Connect(BookReturnListRow)
            borrowedBook={
              Object {
                "borrowedBooks": Array [
                  Object {
                    "actualReturnDate": null,
                    "book": Object {
                      "author": "Aaidh ibn Abdullah al-Qarni",
                      "borrowCount": 6,
                      "id": 2,
                      "quantity": 1,
                      "title": "dont be sad ",
                    },
                    "bookId": 2,
                    "borrowDate": "2018-03-25T16:41:57.255Z",
                    "borrowStatus": "accepted",
                    "expectedReturnDate": "2018-04-08T16:41:57.255Z",
                    "id": 15,
                    "returnStatus": "pending",
                    "user": Object {
                      "id": 2,
                      "username": "mama",
                    },
                    "userId": 2,
                  },
                ],
                "message": "Borrowed books retrieved successfully",
              }
            }
            index={0}
          />
        </tbody>
      </table>
    </div>
  </div>
</div>
`;
